enhancements/cluster-logging/performance-tuning.md

```diff
@@ -0,0 +1,260 @@
+
+``` yaml
+outputs:
+  - name: detailed-tweaks
+    tuning:
+      compression: "zlib"  # Enable protocol-specific compression
+      reconnect:
+        maxDelay: 1s  # Wait max of 1 second between reconnect attempts
+      batch:
+        maxBytes: 50K # Send max 50K of data in each write.
+```
+
+### Goals
+
+- Accommodate users with different performance and reliability trade-offs.
+- Allow users to tune output parameters that we cannot optimize automatically.
+- Expose a simple, general purpose, collector-neutral tuning API
+- Use vector's end-to-end acknowledgements for more efficient reliability.
+
+### Non-Goals
+
+- Not a full end-to-end delivery guarantee, we are limited by reliability of source and sink.
+- Not exposing underlying vector configuration details.
+- No rate limiting - already provided by `outputs[].limit` field.
+- No plans to support fluentd.
+
+## Proposal
+
+### Workflow Description
+
+**logging administrator** is a human user responsible for setting output tuning parameters as needed.
+The logging administrator needs to tune log collection performance and reliability to suit local requirements.
+
+### API Extensions
+
+New field `ClusterLogForwarder.output[].tuning` with the following sub-fields:
+
+#### delivery: AtMostOnce|AtLeastOnce|AtLeastOncePersistent
+
+##### `AtMostOnce`
+
+Logs may be lost if the collector restarts or other faults occur.
+Logs will not be duplicated _by the collector_.
+They may be duplicated in protocol exchanges with the source or sink.
+This mode does no persistent storage and no acknowledgement book-keeping, so has the lowest CPU and disk use.
+
+Use `AtMostOnce` when:
+- It is acceptable to lose some logs due to restarts or faults (network outage, remote store failures)
+- It is important to minimize memory, CPU, disk and network costs.
+- The logging system is expected to run near capacity and is likely to be overloaded.
+
+##### `AtLeastOnce`
+
+Logs _read by the collector_ will not be lost _by the collector_ due to restarts or faults.
+Logs may still be lost
+- Before the collector reads them: the collector cannot keep up with incoming log rate, e.g. log file rotation rate.
+- After the collector sends them: if the output protocol is unreliable, or the target store can drop logs.
+- If rate limiting is enabled: logs may be dropped by the collector to enforce rate limits.
+
+Use `AtLeastOnce` when:
+- It is important to avoid log loss due to restarts or failures.
+- The collector is properly resourced (memory, CPU, disk) for the maximum log throughput at expected peak loads.
+- Log rates are not expected to exceed the collection rate in normal operation.
+
+`AtLeastOnce` uses end-to-end acknowledgements without persistent buffering if possible.
+Acknowledgements give similar reliability to persistent buffering but are more efficient.
+A persistent buffer is used if acknowledgements are not available.
+
+##### `AtLeastOncePersistent`
```

Comment:
Is this something we "may" implement in the future?  I would strongly urge we consider dropping this until we are certain we actually need buffering.  Actual buffering has significant implications for the implementation given we currently (in fluentd) carve out some of the host for our own use.  This may actually require we have some persistent storage.  Consider hiding the text in the rendered output

Reply:
+1 but Note that we already need buffering for cases where acks are not available, which sadly seems to include pod logs :disappointed:  so it's not adding much. I'm OK with dropping this option for the first cut if you think it's not necessary.